activesupport/lib/active_support/callbacks.rb

```diff
@@ -86,21 +88,57 @@ module Callbacks
     #   run_callbacks :save do
     #     save
     #   end
-    def run_callbacks(kind, &block)
-      send "_run_#{kind}_callbacks", &block
-    end
-
-    private
+    #
+    #--
+    #
+    # As this method is used in many places, and often wraps large portions of
+    # user code, it has an additional design goal of minimizing its impact on
+    # the visible call stack. An exception from inside a :before or :after
+    # callback can be as noisy as it likes -- but when control has passed
+    # smoothly through and into the supplied block, we want as little evidence
+    # as possible that we were here.
+    def run_callbacks(kind)
+      callbacks = __callbacks[kind.to_sym]
+
+      if callbacks.empty?
+        yield if block_given?
+      else
+        env = Filters::Environment.new(self, false, nil)
+        next_sequence = callbacks.compile
+
+        invoke_sequence = Proc.new do
+          skipped = nil
+          while true
+            current, next_sequence = next_sequence, next_sequence.nested
+            current.invoke_before(env)
+            if current.final?
+              env.value = !env.halted && (!block_given? || yield)
+            elsif current.skip?(env)
+              (skipped ||= []) << current
+              next
+            else
+              expanded = current.expand_call_template(env, invoke_sequence)
+              expanded.shift.send(*expanded, &expanded.shift)
+            end
+            current.invoke_after(env)
+            skipped.pop.invoke_after(env) while skipped && skipped.first
```

Comment:
```ruby
around_save :foo, if: :no
after_save :bar
```

Both of the above live in the same CallbackSequence. It will be `skipped` because the condition on `foo` is false... but we still have to run `bar`. (Unless we're halted.. but `invoke_after` takes care of that.)

If it was not skipped, then we'd call `foo` and ultimately recurse back into a fresh invocation of `invoke_sequence`, and `bar` would get run by *that* invocation on its way out, as `current`.